stan/math/prim/err/check_corr_matrix.hpp

```diff
@@ -31,28 +32,38 @@ namespace math {
  *   diagonals not near 1, not positive definite, or any of the
  *   elements nan
  */
-template <typename EigMat, require_eigen_matrix_dynamic_t<EigMat>* = nullptr>
+template <typename Mat, require_matrix_t<Mat>* = nullptr>
 inline void check_corr_matrix(const char* function, const char* name,
-                              const EigMat& y) {
-  const auto& y_ref = to_ref(y);
+                              const Mat& y) {
+  auto&& y_ref = to_ref(value_of_rec(y));
   check_square(function, name, y_ref);
   using std::fabs;
   if (y_ref.size() == 0) {
     return;
   }
 
   for (Eigen::Index k = 0; k < y.rows(); ++k) {
-    if (!(fabs(y_ref(k, k) - 1.0) <= CONSTRAINT_TOLERANCE)) {
-      std::ostringstream msg;
-      msg << "is not a valid correlation matrix. " << name << "("
-          << stan::error_index::value + k << "," << stan::error_index::value + k
-          << ") is ";
-      std::string msg_str(msg.str());
-      throw_domain_error(function, name, y_ref(k, k), msg_str.c_str(),
-                         ", but should be near 1.0");
+    if (!(fabs(y_ref.coeff(k, k) - 1.0) <= CONSTRAINT_TOLERANCE)) {
+      [&y_ref, name, k, function]() STAN_COLD_PATH {
```

Comment:
See issue #2249 for more info, but essentially STAN_COLD_PATH is equal to `__attribute__((noinline,cold))` if noinline and cold are available attributes from the compiler. This makes the error checking cheaper since the compiler knows it very rarely has to go down the path of throwing an error and doesn't pay for prefetching the stuff inside of the if().